docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+
+*Example* The following algorithm from [the DOM
+specification](https://dom.spec.whatwg.org) describes how the
+`createCDATASection` method works:
+
+> The `createCDATASection`(*data*) method, when invoked, must run these steps:
+>
+> 1. If [context object](https://dom.spec.whatwg.org/#context-object) is an
+>    [HTML document](https://dom.spec.whatwg.org/#html-document), then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`NotSupportedError`](https://heycam.github.io/webidl/#notsupportederror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 2. If *data* contains the string "`]]>`", then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) an
+>    "[`InvalidCharacterError`](https://heycam.github.io/webidl/#invalidcharactererror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return a new [CDATASection](https://dom.spec.whatwg.org/#cdatasection)
+>    [node](https://dom.spec.whatwg.org/#boundary-point-node) with its
+>    [data](https://dom.spec.whatwg.org/#concept-cd-data) set to *data* and
+>    [node document](https://dom.spec.whatwg.org/#concept-node-document) set to
+>    the [context object](https://dom.spec.whatwg.org/#context-object).
+
+This algorithm begins by ensuring that the context object isn't an HTML
+document. Then, it verifies that the first argument doesn't include a certain
+sequence of characters. Implementations that verify those things in the
+opposite order would achieve a similar result (i.e. they'd reject invalid
+input), but that slight difference could cause interoperability bugs on the
+web.
+
+If we prepare input that violates *both* steps, then we can verify that they're
+taken in the correct order. We can call `createCDATASection` on an HTML
+document and pass the string "`]]>`"; that should produce a
+`NotSupportedError`. If it produces a `InvalidCharacterError` instead, we'll
+know that the implementation has incorrectly placed the second step before the
+first.
+
+## Exercising Restraint
+
+When writing conformance tests, choosing what *not* to test is sometimes just
+as hard as finding what needs testing.
+
+### Skip ambiguity
+
+Specifications occasionally allow browsers some discretion in how they
+implement certain features. These are described using [RFC
+2119](https://tools.ietf.org/html/rfc2119) terms like "MAY" and "OPTIONAL".
+Even though each browser will likely maintain its own tests for the behavior it
+chooses to implement, it's important that shared conformance test suites like
+WPT do not include tests for any particular behavior. If the specification
+doesn't require something, then neither should the tests!
+
+*Example* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.getElementsByTagName`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByTagName):
+
+> The list of elements with qualified name *qualifiedName* for a
```

Comment:
This is a very long algorithm to put into the guide. Could we just say:

_Example_ The algorithm underpinning `document.getElementsByTagName` includes the following paragraph at the end:

> When invoked with the same argument, and as long as root's node document's type has not changed, the same HTMLCollection object may be returned as returned by an earlier call.


I don't think any of the rest of the algorithm is actually relevant to the points made?